ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsController.java

```diff
@@ -140,104 +47,65 @@ public synchronized void onPreviewOrderChanged(StepMovedEvent event) {
      */
     @Subscribe
     public synchronized void onSocketPreviewChanged(SocketPreviewChangedEvent event) {
-        platform.runAsSoonAsPossible(() -> {//Run this function on the main gui thread
-
-            final OutputSocket<?> socket = event.getSocket(); //The socket whose preview has changed
-
-            if (socket.isPreviewed()) {// If the socket was just set as previewed, add it to the list of previewed sockets and add a new view for it.
-
-                if (!this.previewedSockets.contains(socket)) {//If the socket is not already previewed...
-
-                    if (socket.getStep().isPresent()) { //If this is a socket associated with a pipeline step (IE NOT a source)....
-
-                        //Find the appropriate index to add this preview with...
-                        int indexInPreviews = getIndexInPreviewsOfAStepSocket(socket);
-
-                        this.previewedSockets.add(indexInPreviews, socket);//...use this index to add it to the correct location in the list of previews open
-                        this.previewBox.getChildren().add(indexInPreviews, previewViewFactory.create(socket));//...and display it in the correct location in the list of previews open in the gui
-
-                    } else {//This is a socket associated with a source and not a pipeline step...
-
-                        //Find the appropriate index to add this preview with.
-                        int indexInSourcePreviews = getIndexInPreviewsOfASourceSocket(socket);
-
-                        this.previewedSockets.add(indexInSourcePreviews, socket);//Add the preview to the appropriate place in the list of previewed sockets
-                        this.previewBox.getChildren().add(indexInSourcePreviews, previewViewFactory.create(socket));//Display the preview in the appropriate place
-                    }
-                }
-            } else {//The socket was already previewed, so the user must be requesting to not show this preview (remove both it and the corresponding control)
-
-                int index = this.previewedSockets.indexOf(socket);//Get the index of this preview so we can remove the correct entry
-                if (index != -1) {//False when the preview isn't currently displayed
-                    this.previewedSockets.remove(index);
-                    this.eventBus.unregister(this.previewBox.getChildren().remove(index));
-                }
+        OutputSocket<?> socket = event.getSocket();
+
+        @SuppressWarnings("unchecked")
+        ObservableList<SocketPreviewView<?>> previews = (ObservableList) previewBox.getChildren();
+
+        // This needs to run right away to avoid synchronization problems, although in practice this method only runs
+        // in the UI thread anyways (since it fires in response to a button press)
+        PlatformImpl.runAndWait(() -> {
```

Comment:
1. (Currently) this is always run from the UI thread anyways
2. This doesn't have the same deadlock problems that you've identified in some parts of the code base, since the runnable only updates the GUI.  There's only a possibility for deadlocking if we try to update the core in the runnable.
3. This needs to run completely before returning, otherwise we'd have to worry about synchronizing multiple threads sorting the list at the same time, and the UI would also not update instantly